tests/unit/s2n_renegotiate_io_test.c

```diff
@@ -0,0 +1,500 @@
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+            ssize_t app_data_read = 0;
+            uint8_t recv_buffer[S2N_TEST_APP_DATA_LEN * 2] = { 0 };
+            size_t send_i = 0, recv_i = 0;
+
+            /* First handshake */
+            EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+            /* Before Renegotiation */
+            {
+                if (app_data_case & S2N_TEST_APP_DATA_BEFORE_RENEG) {
+                    EXPECT_EQUAL(s2n_send(server_conn, messages[send_i], S2N_TEST_APP_DATA_LEN, &blocked), S2N_TEST_APP_DATA_LEN);
+                    send_i++;
+                }
+
+                EXPECT_SUCCESS(s2n_renegotiate_wipe(server_conn));
+                EXPECT_SUCCESS(s2n_renegotiate_wipe(client_conn));
+            }
+
+            /* Client: ClientHello sent */
+            {
+                int r = s2n_renegotiate(client_conn, recv_buffer, sizeof(recv_buffer), &app_data_read, &blocked);
+                EXPECT_EQUAL(s2n_conn_get_current_message_type(client_conn), SERVER_HELLO);
+                /* Expect to have received any outstanding application data */
+                if (send_i > 0 && recv_i < send_i) {
+                    EXPECT_FAILURE_WITH_ERRNO(r, S2N_ERR_APP_DATA_BLOCKED);
+                    EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_APPLICATION_DATA);
+                    EXPECT_EQUAL(app_data_read, S2N_TEST_APP_DATA_LEN);
+                    EXPECT_BYTEARRAY_EQUAL(recv_buffer, messages[recv_i], S2N_TEST_APP_DATA_LEN);
+                    recv_i++;
+                    reneg_ch_had_app_data = true;
+                } else {
+                    EXPECT_FAILURE_WITH_ERRNO(r, S2N_ERR_IO_BLOCKED);
+                    EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+                    EXPECT_EQUAL(app_data_read, 0);
+                    reneg_ch_had_no_app_data = true;
+                }
+            }
+
+            /* Server: ClientHello recv, ServerHello sent */
+            {
+                if (app_data_case & S2N_TEST_APP_DATA_BEFORE_SERVER_HELLO) {
+                    EXPECT_EQUAL(s2n_send(server_conn, messages[send_i], S2N_TEST_APP_DATA_LEN, &blocked), S2N_TEST_APP_DATA_LEN);
+                    send_i++;
+                }
+
+                EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate(server_conn, &blocked), S2N_ERR_IO_BLOCKED);
+                EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+                EXPECT_NOT_EQUAL(s2n_conn_get_current_message_type(server_conn), SERVER_HELLO);
+                EXPECT_TRUE(IS_NEGOTIATED(server_conn));
+
+                /* All application data sent after the ServerHello is invalid.
+                 * We specifically don't increment sent_i, because we won't use this data.
+                 */
+                if (app_data_case & S2N_TEST_APP_DATA_AFTER_SERVER_HELLO) {
```

Comment:
What is the point of this test case if you never try to read the data that was sent?

Reply:
Actually nvm, I see that w are doing that lower down.